In [12]:
import pandas as pd
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Load Dataset

In [13]:
df = pd.read_excel('../data/raw/online_retail_II.xlsx')

In [18]:
df.shape

(525461, 8)

In [30]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [15]:
df.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='object')

In [16]:
df.dtypes

Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
Price                 float64
Customer ID           float64
Country                object
dtype: object

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB


In [22]:
df.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,525461.000000,525461,525461.000000,417534.000000
mean,10.337667,2010-06-28 11:37:36.845017856,4.688834,15360.645478
min,-9600.000000,2009-12-01 07:45:00,-53594.360000,12346.000000
25%,1.000000,2010-03-21 12:20:00,1.250000,13983.000000
50%,3.000000,2010-07-06 09:51:00,2.100000,15311.000000
75%,10.000000,2010-10-15 12:45:00,4.210000,16799.000000
max,19152.000000,2010-12-09 20:01:00,25111.090000,18287.000000
std,107.424110,NaN,146.126914,1680.811316


In [23]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

In [32]:
print(df["Country"].nunique())

print(df["StockCode"].nunique())

print(df["Invoice"].nunique())

print(df["Customer ID"].nunique())

40
4632
28816
4383


In [33]:
print(df["InvoiceDate"].min())

print(df["InvoiceDate"].max())

2009-12-01 07:45:00
2010-12-09 20:01:00


In [35]:
print(df["Invoice"].astype(str).str.startswith("C").sum())

10206


In [37]:
print((df["Price"] <= 0).sum())

3690


In [38]:
print((df["Quantity"] <= 0).sum())

12326


In [40]:
print(df.duplicated().sum())

6865


In [41]:
df["SalesAmount"] = df["Quantity"] * df["Price"]

df["SalesAmount"].describe()

count    525461.000000
mean         18.154506
std         160.333083
min      -53594.360000
25%           3.750000
50%           9.950000
75%          17.700000
max       25111.090000
Name: SalesAmount, dtype: float64

In [45]:
(df["SalesAmount"] ==0).sum()

np.int64(3687)

In [43]:
df["SalesAmount"].sum()

np.float64(9539484.634)

In [44]:
df.nsmallest(10, "SalesAmount")

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,SalesAmount
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom,-53594.36
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom,-44031.79
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom,-38925.87
241824,C512770,M,Manual,-1,2010-06-17 16:52:00,25111.09,17399.0,United Kingdom,-25111.09
320581,C520667,BANK CHARGES,Bank Charges,-1,2010-08-27 13:42:00,18910.69,NaN,United Kingdom,-18910.69
517953,C537630,AMAZONFEE,AMAZON FEE,-1,2010-12-07 15:04:00,13541.33,NaN,United Kingdom,-13541.33
519294,C537651,AMAZONFEE,AMAZON FEE,-1,2010-12-07 15:49:00,13541.33,NaN,United Kingdom,-13541.33
519170,C537644,AMAZONFEE,AMAZON FEE,-1,2010-12-07 15:34:00,13474.79,NaN,United Kingdom,-13474.79
135012,C502262,M,Manual,-1,2010-03-23 15:20:00,10953.50,12918.0,United Kingdom,-10953.50
135014,C502264,M,Manual,-1,2010-03-23 15:24:00,10953.50,12918.0,United Kingdom,-10953.50


## EDA Summary

### Dataset Overview

- Total Transactions: 525,461
- Customers: 4,383
- Products: 4,632
- Countries: 40
- Date Range: 2009-12-01 to 2010-12-09

### Major Data Quality Issues

- Missing Customer ID: 107,927
- Missing Description: 2,928
- Cancelled Invoices: 10,206
- Quantity <= 0: 12,326
- Price <= 0: 3,690
- Duplicate Rows: 6,865

### Candidate Dimensions

- Customer
- Product
- Country
- Time

### Candidate Measures

- Quantity
- Price
- SalesAmount

### Source Data Grain

One row represents one product purchased within one invoice.